README.md
# API Auth
BADGES

API Auth provides an authentication system for API and SPA.
It uses a PSR-15 middleware and parts to generate token for login requests are provided.

## How it works

First, your user get a token from an endpoint of your API (let's say ```/token``` endpoint using POST request).
If this request come from a known domain (for example, your SPA), a cookie is set to client browser with expiration of 30 minutes.
Otherwise, no cookie is set and token is only return by response body as JSON.
Body response is like: ```{"token": "MY-TOKEN"}```.

### Requests with cookie

Cookie token is secured and inaccessible from client.
Each time a request is made to the API, if token is still valid, expiration is set to another 30 minutes.
Session will expire after 30 minutes without request.

### Requests without cookie

If no cookie have been set from your login endpoint, it means it is not an allowed domain (for example, a mobile application).
Token should have been stored to a secure storage.
Each time you make a request, you need to authenticate with Bearer authorization header.  
```Authorization: Bearer MY-TOKEN```  

### No authentication

If a request is made without cookie, without Authorization request, or with a bad token, ```AuthMiddleware``` throw an ```AuthException```.
You should catch this exception to return an appropriate response to your client (and redirect to login page).

## Installation

It is recommended to install this package using [Composer](https://getcomposer.org/)

```shell
composer require t0mmy742/api-auth
```

API Auth requires PHP 8.0 or newer

## Usage

This package is composed of two parts: token creation and token validation.

### Create authentication token

You probably want to authenticate your users using the ```/token``` or ```/login``` endpoint of your API.
Your endpoint can have the following Controller:

```php
<?php

declare(strict_types=1);

namespace YOUR\NAMESPACE\Controller;

use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use T0mmy742\ApiAuthentication\Token\Generation\TokenGenerationTrait;

use function is_array;
use function json_encode;

class TokenController
{
    use TokenGenerationTrait;

    /** @var string[] */
    private array $cookieDomains = ['example.org'];

    public function __invoke(ServerRequestInterface $request, ResponseInterface $response): ResponseInterface
    {
        $params = $request->getParsedBody();

        if (is_array($params)) {
            $login = $params['login'] ?? null;
            $pass = $params['pass'] ?? null;

            if ($login !== null && $pass !== null) {
                if ($this->validAuth($login, $pass)) {
                    // Create a new token
                    $token = $this->createToken();
                    // Add token to a cookie if origin domain is allowed ($this->cookieDomains). Can be from a SPA
                    $response = $this->manageCookieResponse($request, $response, $token);
                }
            }
        }

        // Write token to the response body
        $response->getBody()->write(json_encode(['token' => $token]));
        
        return $response;
    }

    private function validAuth(string $login, string $pass): bool
    {
        // You probably want to check in database...
        return true;
    }
}

```

### Validate authentication token

To validate token, you need to add ```AuthMiddleware``` class to your middleware stack.
This middleware has the following signature:
```php
public function __construct(\T0mmy742\ApiAuthentication\Token\Validation\TokenValidationInterface $tokenValidation, array $cookieDomains = []);
```
```$cookieDomains``` is an array containing allowed domains to set a cookie named ```access_token``` with the authentication token.
It is useful to authenticate a user from your SPA.
Mobile application does not have domain, and use Bearer token authentication.  
For example, if your API host is ```api.example.org```, your SPA is probably ```example.org```. Then, we have ```$cookieDomains = ['example.org']```.

```$tokenValidation``` is a class implementing ```TokenValidationInterface```.
Implementation can be as bellow:
```php
<?php

declare(strict_types=1);

namespace YOUR\NAMESPACE\Auth;

use T0mmy742\ApiAuthentication\Token\Validation\TokenValidationTrait;

class TokenValidation implements \T0mmy742\ApiAuthentication\Token\Validation\TokenValidationInterface
{
    use TokenValidationTrait;

    private function retrieveUserId(string $token): ?string
    {
        // You probably want to make request to your database to get userId from this token
        // Return userId if exists, return null otherwise.
        return '1'
    }
}

```

## License

API Auth is licensed under the [MIT license](LICENSE.md).

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added


tests/Token/TokenCookieTraitTest.php
<?php

declare(strict_types=1);

namespace T0mmy742\ApiAuthentication\Tests\Token;

use PHPUnit\Framework\TestCase;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use T0mmy742\ApiAuthentication\Token\TokenCookieTrait;

use function gmdate;
use function time;

/**
 * @covers \T0mmy742\ApiAuthentication\Token\TokenCookieTrait
 */
class TokenCookieTraitTest extends TestCase
{
    use TokenCookieTrait;

    /** @var string[] */
    private array $cookieDomains = ['example.org'];

    public function testAdd30MinCookieToResponse(): void
    {
        $token = '8ba28a85ec598409d31ffc5e21117d0973f83040fac7599fbfd2db8edab84705';
        $domain = 'example.org';

        $response = $this->createMock(ResponseInterface::class);
        $response2 = $this->createStub(ResponseInterface::class);
        $response
            ->expects(self::once())
            ->method('withAddedHeader')
            ->with(self::isType('string'))
            ->willReturnCallback(
                function (string $name, string $value) use ($token, $domain, $response2): ResponseInterface {
                    $expires = gmdate('D, d-M-Y H:i:s e', time() + 30 * 60);
                    $expectedCookie = 'access_token=' . $token
                        . '; Expires=' . $expires
                        . '; Domain=' . $domain
                        . '; Path=' . '/'
                        . '; Secure'
                        . '; HttpOnly'
                        . '; SameSite=Strict';

                    self::assertSame($expectedCookie, $value);

                    return $response2;
                }
            );

        $newResponse = $this->add30MinCookieToResponse($response, $token, $domain);

        self::assertSame($response2, $newResponse);
    }

    public function testFromAllowedDomainRefererHeader(): void
    {
        $expectedDomain = 'example.org';
        $uri = 'https://' . $expectedDomain . '/endpoint';

        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::once())
            ->method('hasHeader')
            ->with('Referer')
            ->willReturn(true);
        $request
            ->expects(self::once())
            ->method('getHeader')
            ->with('Referer')
            ->willReturn([$uri]);

        $domain = $this->fromAllowedDomains($request);

        self::assertSame($expectedDomain, $domain);
    }

    public function testFromAllowedDomainOriginHeader(): void
    {
        $expectedDomain = 'example.org';
        $uri = 'https://' . $expectedDomain . '/endpoint';

        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::exactly(2))
            ->method('hasHeader')
            ->withConsecutive(['Referer'], ['Origin'])
            ->willReturnOnConsecutiveCalls(false, true);
        $request
            ->expects(self::once())
            ->method('getHeader')
            ->with('Origin')
            ->willReturn([$uri]);

        $domain = $this->fromAllowedDomains($request);

        self::assertSame($expectedDomain, $domain);
    }

    public function testFromAllowedDomainBadHeader(): void
    {
        $badUri = 'https:/e^f';

        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::once())
            ->method('hasHeader')
            ->with('Referer')
            ->willReturn(true);
        $request
            ->expects(self::once())
            ->method('getHeader')
            ->with('Referer')
            ->willReturn([$badUri]);

        $domain = $this->fromAllowedDomains($request);

        self::assertNull($domain);
    }

    public function testFromAllowedDomainNoHeader(): void
    {
        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::exactly(2))
            ->method('hasHeader')
            ->withConsecutive(['Referer'], ['Origin'])
            ->willReturnOnConsecutiveCalls(false);

        $domain = $this->fromAllowedDomains($request);

        self::assertNull($domain);
    }

    public function testFromAllowedDomainNotAllowed(): void
    {
        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::once())
            ->method('hasHeader')
            ->with('Referer')
            ->willReturn(true);
        $request
            ->expects(self::once())
            ->method('getHeader')
            ->with('Referer')
            ->willReturn(['https://badexample.org/endpoint']);

        $domain = $this->fromAllowedDomains($request);

        self::assertNull($domain);
    }
}


tests/Middleware/AuthMiddlewareTest.php
<?php

declare(strict_types=1);

namespace T0mmy742\ApiAuthentication\Tests\Middleware;

use PHPUnit\Framework\TestCase;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\RequestHandlerInterface;
use T0mmy742\ApiAuthentication\Exceptions\AuthException;
use T0mmy742\ApiAuthentication\Middleware\AuthMiddleware;
use T0mmy742\ApiAuthentication\Token\Validation\TokenValidationInterface;

/**
 * @covers \T0mmy742\ApiAuthentication\Middleware\AuthMiddleware
 */
class AuthMiddlewareTest extends TestCase
{
    public function testFromAllowedDomainsNoCookie(): void
    {
        $cookieDomain = 'example.org';

        $tokenValidation = $this->createStub(TokenValidationInterface::class);

        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::once())
            ->method('hasHeader')
            ->willReturn(true);
        $request
            ->expects(self::once())
            ->method('getHeader')
            ->willReturn(['https://' . $cookieDomain]);
        $request
            ->expects(self::once())
            ->method('getCookieParams')
            ->willReturn([]);

        $handler = $this->createStub(RequestHandlerInterface::class);

        $this->expectException(AuthException::class);
        $this->expectExceptionMessage('No token found on request');
        (new AuthMiddleware($tokenValidation, [$cookieDomain]))->process($request, $handler);
    }

    public function testFromAllowedDomainsCookieInvalidToken(): void
    {
        $cookieDomain = 'example.org';
        $token = '8ba28a85ec598409d31ffc5e21117d0973f83040fac7599fbfd2db8edab84705';

        $tokenValidation = $this->createMock(TokenValidationInterface::class);

        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::once())
            ->method('hasHeader')
            ->willReturn(true);
        $request
            ->expects(self::once())
            ->method('getHeader')
            ->willReturn(['https://' . $cookieDomain]);
        $request
            ->expects(self::once())
            ->method('getCookieParams')
            ->willReturn(['access_token' => $token]);

        $tokenValidation
            ->expects(self::once())
            ->method('validateToken')
            ->with($token, $request)
            ->willThrowException(new AuthException());

        $handler = $this->createStub(RequestHandlerInterface::class);

        $this->expectException(AuthException::class);
        (new AuthMiddleware($tokenValidation, [$cookieDomain]))->process($request, $handler);
    }

    public function testFromAllowedDomainsCookieValidToken(): void
    {
        $cookieDomain = 'example.org';
        $token = '8ba28a85ec598409d31ffc5e21117d0973f83040fac7599fbfd2db8edab84705';

        $tokenValidation = $this->createMock(TokenValidationInterface::class);

        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::once())
            ->method('hasHeader')
            ->willReturn(true);
        $request
            ->expects(self::once())
            ->method('getHeader')
            ->willReturn(['https://' . $cookieDomain]);
        $request
            ->expects(self::once())
            ->method('getCookieParams')
            ->willReturn(['access_token' => $token]);

        $tokenValidation
            ->expects(self::once())
            ->method('validateToken')
            ->with($token, $request)
            ->willReturn($request);

        $response = $this->createMock(ResponseInterface::class);
        $response
            ->expects(self::once())
            ->method('withAddedHeader')
            ->with('Set-Cookie', self::isType('string'))
            ->willReturnSelf();

        $handler = $this->createMock(RequestHandlerInterface::class);
        $handler
            ->expects(self::once())
            ->method('handle')
            ->with($request)
            ->willReturn($response);

        (new AuthMiddleware($tokenValidation, [$cookieDomain]))->process($request, $handler);
    }

    public function testNotFromAllowedDomainsCookieNoTokenHeader(): void
    {
        $cookieDomain = 'example.org';

        $tokenValidation = $this->createStub(TokenValidationInterface::class);

        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::exactly(2))
            ->method('hasHeader')
            ->willReturn(false);
        $request
            ->expects(self::once())
            ->method('getHeader')
            ->with('Authorization')
            ->willReturn([]);

        $handler = $this->createStub(RequestHandlerInterface::class);

        $this->expectException(AuthException::class);
        $this->expectExceptionMessage('No token found on request');
        (new AuthMiddleware($tokenValidation, [$cookieDomain]))->process($request, $handler);
    }

    public function testNotFromAllowedDomainsCookieNotABearerTokenHeader(): void
    {
        $cookieDomain = 'example.org';
        $token = '8ba28a85ec598409d31ffc5e21117d0973f83040fac7599fbfd2db8edab84705';

        $tokenValidation = $this->createStub(TokenValidationInterface::class);

        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::exactly(2))
            ->method('hasHeader')
            ->willReturn(false);
        $request
            ->expects(self::once())
            ->method('getHeader')
            ->with('Authorization')
            ->willReturn(['NOTBEARER ' . $token]);

        $handler = $this->createStub(RequestHandlerInterface::class);

        $this->expectException(AuthException::class);
        $this->expectExceptionMessage('No Bearer token found on Authorization header');
        (new AuthMiddleware($tokenValidation, [$cookieDomain]))->process($request, $handler);
    }

    public function testNotFromAllowedDomainsCookieBearerTokenHeaderBadLength(): void
    {
        $cookieDomain = 'example.org';
        $token = 'BAD-TOKEN-INVALID-LENGTH';

        $tokenValidation = $this->createStub(TokenValidationInterface::class);

        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::exactly(2))
            ->method('hasHeader')
            ->willReturn(false);
        $request
            ->expects(self::once())
            ->method('getHeader')
            ->with('Authorization')
            ->willReturn(['Bearer ' . $token]);

        $handler = $this->createStub(RequestHandlerInterface::class);

        $this->expectException(AuthException::class);
        $this->expectExceptionMessage('Bad Bearer token found on Authorization header');
        (new AuthMiddleware($tokenValidation, [$cookieDomain]))->process($request, $handler);
    }

    public function testNotFromAllowedDomainsCookieBearerTokenHeaderInvalid(): void
    {
        $cookieDomain = 'example.org';
        $token = '8ba28a85ec598409d31ffc5e21117d0973f83040fac7599fbfd2db8edab84705';

        $tokenValidation = $this->createMock(TokenValidationInterface::class);

        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::exactly(2))
            ->method('hasHeader')
            ->willReturn(false);
        $request
            ->expects(self::once())
            ->method('getHeader')
            ->with('Authorization')
            ->willReturn(['Bearer ' . $token]);

        $tokenValidation
            ->expects(self::once())
            ->method('validateToken')
            ->with($token, $request)
            ->willThrowException(new AuthException());

        $handler = $this->createStub(RequestHandlerInterface::class);

        $this->expectException(AuthException::class);
        (new AuthMiddleware($tokenValidation, [$cookieDomain]))->process($request, $handler);
    }

    public function testNotFromAllowedDomainsCookieBearerTokenHeaderValid(): void
    {
        $cookieDomain = 'example.org';
        $token = '8ba28a85ec598409d31ffc5e21117d0973f83040fac7599fbfd2db8edab84705';

        $tokenValidation = $this->createMock(TokenValidationInterface::class);

        $request = $this->createMock(ServerRequestInterface::class);
        $request
            ->expects(self::exactly(2))
            ->method('hasHeader')
            ->willReturn(false);
        $request
            ->expects(self::once())
            ->method('getHeader')
            ->with('Authorization')
            ->willReturn(['Bearer ' . $token]);

        $tokenValidation
            ->expects(self::once())
            ->method('validateToken')
            ->with($token, $request)
            ->willReturn($request);

        $response = $this->createStub(ResponseInterface::class);

        $handler = $this->createMock(RequestHandlerInterface::class);
        $handler
            ->expects(self::once())
            ->method('handle')
            ->with($request)
            ->willReturn($response);

        (new AuthMiddleware($tokenValidation, [$cookieDomain]))->process($request, $handler);
    }
}


tests/Token/Validation/TokenValidationHelper.php
<?php

declare(strict_types=1);

namespace T0mmy742\ApiAuthentication\Tests\Token\Validation;

use T0mmy742\ApiAuthentication\Token\Validation\TokenValidationTrait;

class TokenValidationHelper// implements \T0mmy742\ApiAuthentication\Token\Validation\TokenValidationInterface
{
    use TokenValidationTrait;

    private bool $tokenValid;

    public function __construct(bool $tokenValid)
    {
        $this->tokenValid = $tokenValid;
    }

    private function retrieveUserId(string $token): ?string
    {
        if ($this->tokenValid) {
            return '1';
        } else {
            return null;
        }
    }
}


src/Middleware/AuthMiddleware.php
<?php

declare(strict_types=1);

namespace T0mmy742\ApiAuthentication\Middleware;

use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Http\Server\MiddlewareInterface;
use Psr\Http\Server\RequestHandlerInterface;
use T0mmy742\ApiAuthentication\Exceptions\AuthException;
use T0mmy742\ApiAuthentication\Token\TokenCookieTrait;
use T0mmy742\ApiAuthentication\Token\Validation\TokenValidationInterface;

use function str_starts_with;
use function strlen;
use function substr;

class AuthMiddleware implements MiddlewareInterface
{
    use TokenCookieTrait;

    private TokenValidationInterface $tokenValidation;
    /** @var string[] */
    private array $cookieDomains;

    /**
     * @param TokenValidationInterface $tokenValidation
     * @param string[] $cookieDomains
     */
    public function __construct(TokenValidationInterface $tokenValidation, array $cookieDomains = [])
    {
        $this->tokenValidation = $tokenValidation;
        $this->cookieDomains = $cookieDomains;
    }

    /**
     * Process the token validation. It retrieves token from either a cookie or the Authorization header,
     * then validates it. If token is invalid or not present, AuthException is thrown.
     *
     * @param ServerRequestInterface $request
     * @param RequestHandlerInterface $handler
     * @return ResponseInterface
     * @throws AuthException if token is not present or invalid.
     *
     * @uses TokenCookieTrait::fromAllowedDomains()
     */
    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
    {
        $domain = $this->fromAllowedDomains($request);
        if ($domain !== null) {
            $token = $this->retrieveTokenFromCookie($request);
            $request = $this->tokenValidation->validateToken($token, $request);

            return $this->add30MinCookieToResponse($handler->handle($request), $token, $domain);
        } else {
            $token = $this->retrieveTokenFromHeader($request);
            $request = $this->tokenValidation->validateToken($token, $request);

            return $handler->handle($request);
        }
    }

    /**
     * Retrieve token from a cookie. Throws AuthException if no token cookie is present.
     *
     * @param ServerRequestInterface $request
     * @return string the token present in cookie.
     * @throws AuthException if token is not present.
     */
    private function retrieveTokenFromCookie(ServerRequestInterface $request): string
    {
        $token = $request->getCookieParams()['access_token'] ?? null;
        if ($token === null) {
            throw new AuthException('No token found on request');
        }

        return $token;
    }

    /**
     * Retrieve token from the Authorization header. Throws AuthException if no token present or invalid.
     *
     * @param ServerRequestInterface $request
     * @return string the token present in Authorization header.
     * @throws AuthException if token is not present or invalid.
     */
    private function retrieveTokenFromHeader(ServerRequestInterface $request): string
    {
        $authorizationHeader = $request->getHeader('Authorization')[0] ?? null;
        if ($authorizationHeader === null) {
            throw new AuthException('No token found on request');
        }
        if (!str_starts_with($authorizationHeader, 'Bearer ')) {
            throw new AuthException('No Bearer token found on Authorization header');
        }
        $token = substr($authorizationHeader, 7);
        if (strlen($token) !== 64) {
            throw new AuthException('Bad Bearer token found on Authorization header');
        }

        return $token;
    }
}


src/Token/Generation/TokenGenerationTrait.php
<?php

declare(strict_types=1);

namespace T0mmy742\ApiAuthentication\Token\Generation;

use Exception;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;
use T0mmy742\ApiAuthentication\Token\TokenCookieTrait;

use function bin2hex;
use function random_bytes;

trait TokenGenerationTrait
{
    use TokenCookieTrait;

    /**
     * Create a token based on {@link random_bytes()}.
     *
     * @return string a newly generated token.
     * @throws Exception if an appropriate source of randomness cannot be found.
     */
    private function createToken(): string
    {
        return bin2hex(random_bytes(32));
    }

    /**
     * Define Set-Cookie to response header if necessary.
     *
     * @param ServerRequestInterface $request the server request.
     * @param ResponseInterface $response the response.
     * @param string $token the generated token.
     * @return ResponseInterface the response with cookie header if needed.
     *
     * @uses TokenCookieTrait::fromAllowedDomains()
     * @uses TokenCookieTrait::add30MinCookieToResponse()
     */
    private function manageCookieResponse(
        ServerRequestInterface $request,
        ResponseInterface $response,
        string $token
    ): ResponseInterface {
        $domain = $this->fromAllowedDomains($request);

        if ($domain !== null) {
            $response = $this->add30MinCookieToResponse($response, $token, $domain);
        }

        return $response;
    }
}


src/Token/Validation/TokenValidationTrait.php
<?php

declare(strict_types=1);

namespace T0mmy742\ApiAuthentication\Token\Validation;

use Psr\Http\Message\ServerRequestInterface;
use T0mmy742\ApiAuthentication\Exceptions\AuthException;

trait TokenValidationTrait
{
    /**
     * @see TokenValidationInterface::validateToken()
     */
    public function validateToken(string $token, ServerRequestInterface $request): ServerRequestInterface
    {
        $userId = $this->retrieveUserId($token);

        if ($userId === null) {
            throw new AuthException('Token is not valid');
        }

        return $request->withAttribute('uid', $userId);
    }

    /**
     * Check if received token is valid. It should return user ID corresponding to this token. When null is returned
     * value, it means no user match this token.
     * This function can be used to update a 'last_access' column of the database table of this token.
     *
     * @param string $token
     * @return string|null The user ID corresponding to this token, null if token is invalid.
     */
    abstract private function retrieveUserId(string $token): ?string;
}


src/Token/Validation/TokenValidationInterface.php
<?php

declare(strict_types=1);

namespace T0mmy742\ApiAuthentication\Token\Validation;

use Psr\Http\Message\ServerRequestInterface;
use T0mmy742\ApiAuthentication\Exceptions\AuthException;

interface TokenValidationInterface
{
    /**
     * Check if received token is valid. It should return a new ServerRequestInterface with an attribute having the
     * user ID corresponding to this token. If no user match this token, AuthException thrown.
     *
     * @param string $token the token to validate.
     * @param ServerRequestInterface $request the server request from where come the token.
     * @return ServerRequestInterface A new server request with the user ID corresponding to this token as an attribute,
     * null if token is invalid.
     * @throws AuthException if no user match the token
     */
    public function validateToken(string $token, ServerRequestInterface $request): ServerRequestInterface;
}


src/Token/TokenCookieTrait.php
<?php

declare(strict_types=1);

namespace T0mmy742\ApiAuthentication\Token;

use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\ServerRequestInterface;

use function in_array;
use function is_string;
use function parse_url;

use const PHP_URL_HOST;

trait TokenCookieTrait
{
    /**
     * Add a cookie value expiring in 30 minutes from now to the response.
     *
     * @param ResponseInterface $response
     * @param string $token
     * @param string $domain
     * @return ResponseInterface
     *
     * @uses TokenCookieTrait::generate30MinCookie()
     */
    private function add30MinCookieToResponse(
        ResponseInterface $response,
        string $token,
        string $domain
    ): ResponseInterface {
        return $response->withAddedHeader(
            'Set-Cookie',
            $this->generate30MinCookie($token, $domain)
        );
    }

    /**
     * Generate a cookie value expiring in 30 minutes from now.
     *
     * @param string $token
     * @param string $domain
     * @return string
     *
     * @uses TokenCookieTrait::generateCookie()
     */
    private function generate30MinCookie(string $token, string $domain): string
    {
        return $this->generateCookie('access_token', $token, gmdate('D, d-M-Y H:i:s e', time() + 30 * 60), $domain);
    }

    /**
     * Generate a cookie value of Set-Cookie header based on parameters.
     *
     * @param string $cookieName
     * @param string $token
     * @param string $expires
     * @param string $domain
     * @return string
     */
    private function generateCookie(string $cookieName, string $token, string $expires, string $domain): string
    {
        return $cookieName . '=' . $token
            . '; Expires=' . $expires
            . '; Domain=' . $domain
            . '; Path=' . '/'
            . '; Secure'
            . '; HttpOnly'
            . '; SameSite=Strict';
    }

    /**
     * Check if domain of the current request is part of $this->cookieDomains.
     *
     * @param ServerRequestInterface $request
     * @return string|null
     */
    private function fromAllowedDomains(ServerRequestInterface $request): ?string
    {
        if ($request->hasHeader('Referer')) {
            $uriHeader = $request->getHeader('Referer')[0];
        } elseif ($request->hasHeader('Origin')) {
            $uriHeader = $request->getHeader('Origin')[0];
        } else {
            return null;
        }

        $domain = parse_url($uriHeader, PHP_URL_HOST);

        if (!is_string($domain)) {
            return null;
        }

        return in_array($domain, $this->cookieDomains, true) ? $domain : null;
    }
}
